### PR TITLE
Fluent style builder API support(#3431)

### PR DESCRIPTION
**Pull request by Bricks-Man**
## What is the purpose of the change

Fluent style builder API support

## Brief changelog

see [issue](https://github.com/apache/incubator-dubbo/issues/3431)

## Verifying this change

see UT

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by chickenlj** [comment]
Awesome!

Looks like we still lack a unified entrance to bootstrap the `refer` and `export` process of all services and references.

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/annotation/Method.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.apache.dubbo.common.Constants;
+
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+@Inherited
+public @interface Method {
```
[the remark this answers is not in the document]
Comment (Bricks-Man) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/annotation/Method.java]: Oh!!!  It's my mistake.I found there is a xml config named <dubbo:method> but can't found Method annotation.I want to add one to cover it.But I mistakenly submitted it to this pr.Please forgive my foolishness.

[the remark this answers is not in the document]
Comment (chickenlj) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/annotation/Method.java]: Would you please remove the Method annotation in this PR if it has nothing to do with the target issue?

As for method annotation, I noticed one effort from cvictory, please feel free to share your opinions on it https://github.com/apache/incubator-dubbo/pull/2603

[the remark this answers is not in the document]
Comment (Bricks-Man) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/annotation/Method.java]: Ok! I will remot it,as soon as possible!
And I found that my pr couldn't pass the check.Could you help me?When I ran the command ‘`mvn clean install -DskipTests=false`’, I passed it locally.

**Comment by chickenlj** [comment]
Let's put this PR under milestone `2.7.2`.

**Comment by ralf0131** [comment]
@chickenlj What is your conclusion on this pull request? Request change or approve? It is better to follow the [official review process](https://help.github.com/en/articles/approving-a-pull-request-with-required-reviews)

**Comment by kezhenxu94** [comment]
@Bricks-Man Thanks for your contribution

**Comment by carryxyh** [comment]
I don't think we need so many builders like `AbstractInterfaceBuilder`. It seems that this class only provides some public properties. And users don't need to use these builders at all, even inside dubbo.

I think we can use the builder to inherit existing classes to do the reuse of using public properties, instead of turning all Config into builders.

IMO, fluent api is just for users, we can make code like this:

```
AbstractReferenceConfig {
	
    protected Boolean check = true;

    public void setCheck(Boolean check) {
        this.check = check;
    }

    public Boolean isCheck() {
    	return this.check;
    }
}

ReferenceBuilder extends AbstractReferenceConfig {
	
    public ReferenceBuilder setCheck(Boolean check) {
        super.setCheck(check);
        return this;
    }

    public ReferenceConfig build() {
    	ReferenceConfig r = new ReferenceConfig();
    	r.setCheck(check);
    	return r;
    }
}
```

What do you think?

**Comment by Bricks-Man** [comment]
Without some abstract builder like `AbstractInterfaceBuilder`,i need many duplicate code to provides public properties in builders like `ServiceBuilder` and `ProviderBuilder`.Do you have any ideas for this?

And I think your builder code is very clean.My code is a bit redundant.My code only covers all previous test cases.It's worked,and just only worked.These builders are not conducive to modification,when Config change.I will change it.Thank you!

**Comment by carryxyh** [comment]
Hi, @Bricks-Man 

`Without some abstract builder like AbstractInterfaceBuilder,i need many duplicate code to provides public properties in builders like ServiceBuilder and ProviderBuilder.Do you have any ideas for this?`

It seems that we can't solve the problem of duplicate code. In your scenario, duplicate code is placed in the builder. For example, the `setCheck` code appears in the builder and origin config. In my scenario, `setCheck` is still repeated in builder. Since neither can solve the problem of duplicate code, we can try to reduce the number of classes. 
:)

**Comment by chickenlj** [comment]
Do you all think this PR has been ready to be released in 2.7.1? I would suggest putting it into 2.7.2 it's not been fully discussed yet.

**Comment by Bricks-Man** [comment]
@carryxyh IMO,i think `AbstractInterfaceBuilder` can be extend like `AbstractInterface` by developers,so i add some abstract builder.And now I think builder is for user,not for developer,after i removed abstract builder.Devoper will find there is  a `AbstractInterface`, but there is no `AbstractInterfaceBuilder`.

@chickenlj How about your opion for this?

**Comment by carryxyh** [comment]
Hi, @Bricks-Man 
I think there is no problem with this. `Abstractxxxconfig` does not need to correspond to a `Abstractxxxbuilder`.

@chickenlj 
I think this pr was merged a bit early, because there is quite a bit of code waiting for us to review. But I found this pr to be marked with the `READY-TO-MERGE`.

**Comment by carryxyh** [comment]
Should we revert this pr first. 
It seems that @Bricks-Man still has some places to modify, what do you think?

**Comment by beiwei30** [comment]
looks like we revisit this merge in 2.7.2, this change looks big.

**Comment by cvictory** [comment]
> looks like we revisit this merge in 2.7.2, this change looks big.

we can revisit this after some bootstrap was defined by @mercyblitz

**Pull request merged**